File: lib/superdirective_filter.py
import numpy as np
from scipy.spatial.distance import pdist, squareform


def superdirective_filter(micPosition, thetaAx, fAx, c):
    fLen = len(fAx)
    micsN = micPosition.shape[0]

    dl = 10**(-20/10)  # Diagonal loading
    gSteeringAll = np.zeros((micsN, len(thetaAx), fLen), dtype=complex)

    k = np.column_stack((np.cos(thetaAx), np.sin(thetaAx)))
    for mm in range(micsN):
        for aa in range(len(thetaAx)):
            gSteeringAll[mm, aa, :] = np.exp(1j * 2 * np.pi * fAx / c * np.dot(k[aa, :], micPosition[mm, :]))

    dist = squareform(pdist(micPosition))
    Gamma2 = np.zeros((micsN*micsN, fLen))

    i = 0
    for rw in range(micsN):
        for co in range(micsN):
            Gamma2[i, :] = np.sinc(2 * np.pi * fAx * dist[rw, co] / (np.pi * c))
            i += 1

    GammaRS = Gamma2.reshape(micsN, micsN, fLen)

    eye = np.eye(micsN)
    long_eye = np.zeros((micsN, micsN, fLen))
    for ff in range(fLen):
        long_eye[:, :, ff] = eye

    GammaRS = GammaRS + dl * np.max(np.max(GammaRS)) * long_eye

    h = np.zeros((micsN, len(thetaAx), fLen), dtype=complex)

    for ff in range(fLen):
        gSt = gSteeringAll[:, :, ff]
        gamma = GammaRS[:, :, ff]
        DD = np.diag(np.dot(np.conj(gSt.T), np.linalg.solve(gamma, gSt)))
        h[:, :, ff] = np.linalg.solve(gamma, gSt) / DD

    H = np.conj(h)

    return np.transpose(H, (0, 2, 1))

File: parametric_vm/lib/superdirectivefilter.m
function H = superdirectivefilter(micPosition, thetaAx, fAx, c)
    %% PSEUDOSPECTRUM
    % This function computes the filter for an array over 360
    % degrees with superdirective beamforming algorithm
    %
    % Params:
    %   micPosition: the relative microphone position
    %   c: sound speed
    % Returns:
    %   H: the filter in frequency

    fLen = length(fAx);
    micPosition = double(micPosition);
    micsN = size(micPosition, 1);
    c = double(c);
    dl = 10^(-20/10);  % Diagonal loading

    % Filter definition
    gSteeringAll = zeros(micsN, length(thetaAx), fLen);     % Steering Vector
    k = [cos(thetaAx)', sin(thetaAx)'];
    for mm = 1:micsN
        for aa = 1:length(thetaAx)
            gSteeringAll(mm, aa, :) = exp(1i*2*pi*fAx/c .* (k(aa,:) * micPosition(mm,:)')).';
        end
    end

    dist = squareform(pdist(micPosition));                  % Microphone dist
    Gamma2 = zeros(micsN*micsN, fLen);                      % Cross-correlation

    for mm = 1:micsN*micsN
        Gamma2(mm,:) = sinc(2*pi*fAx * dist(mm) / (pi*c));
    end

    GammaRS = reshape(Gamma2, micsN, micsN, fLen);
    GammaRS = GammaRS + dl * max(max(GammaRS)) .* repmat(eye(micsN), 1,1, fLen);

    for ff = 1:fLen
       gSt = gSteeringAll(:,:,ff);                      % Steeering vector
       gamma = GammaRS(:,:,ff);                         % Cross-correlation
       DD = diag(gSt' * (gamma \ gSt));                 % Denominator
       h(:,:,ff) =  (gamma \ gSt) ./ DD';               % Filter!
    end
    H = conj(h);

end
